stdlib/public/core/StringBridge.swift

```diff
@@ -337,42 +364,75 @@ internal func _cocoaASCIIPointer(_ str: _CocoaString) -> UnsafePointer<UInt8>? {
   return _NSStringASCIIPointer(_objc(str))
 }
 
-private enum CocoaStringPointer {
-  case ascii(UnsafePointer<UInt8>)
-  case utf8(UnsafePointer<UInt8>)
-  case utf16(UnsafePointer<UInt16>)
-  case none
+// This does not create a bridged String that's safe to use past the current
+// stack frame, e.g. it does not defensively copy it and so on. Returns true
+// if we were able to quickly bridge a String, or false if not
+internal func _withTemporaryBridgedCocoaString(
+  _ cocoaString: _CocoaString,
+  _ work: (String) -> Void
+) -> Bool {
+  switch _KnownCocoaString(cocoaString) {
+  case .storage(let storage):
+    work(storage.asString)
+    return true
+  case .shared(let storage):
+    work(storage.asString)
+    return true
+    #if !(arch(i386) || arch(arm))
+  case .tagged(let str):
+    let tmpStr = String(_StringGuts(_SmallString(taggedCocoa: str)))
+    work(tmpStr)
+    return true
+    #endif
+  case .mutable(let storage):
+    work(storage.asString)
+    return true
+  case .cocoa(let str):
+    let length = _stdlib_binary_CFStringGetLength(str)
+    let guts:_StringGuts
+    if let ascii = _cocoaASCIIPointer(str) {
+      // form an immortal contiguous-ASCII string if we can
+      let asciiBuffer = UnsafeBufferPointer(start: ascii, count: length)
+      guts = _StringGuts(asciiBuffer, isASCII: true)
+      work(String(guts))
+      return true
+    }
+    return false
+  }
 }
 
-@_effects(readonly)
-private func _getCocoaStringPointer(
-  _ cfImmutableValue: _CocoaString
-) -> CocoaStringPointer {
-  if let asciiPtr = _cocoaASCIIPointer(cfImmutableValue) {
-    // NOTE: CFStringGetCStringPointer means ASCII
-    return .ascii(asciiPtr)
-  }
-  if let utf16Ptr = _stdlib_binary_CFStringGetCharactersPtr(cfImmutableValue) {
-    return .utf16(utf16Ptr)
-  }
-  return .none
+internal func _withCocoaStringUTF16Contents(
+  _ cocoaString: _CocoaString,
+  _ work: (UnsafeBufferPointer<UTF16.CodeUnit>) -> Void
+) {
+  let length = _stdlib_binary_CFStringGetLength(cocoaString)
+  let buffer = UnsafeMutableBufferPointer<UTF16.CodeUnit>.allocate(
+    capacity: length
+  )
+  _cocoaStringCopyCharacters(
+    from: cocoaString,
+    range: 0 ..< length,
+    into: buffer.baseAddress!
+  )
+  work(UnsafeBufferPointer(buffer))
+  buffer.deallocate()
```

Comment:
Err, don't we want to check `_stdlib_binary_CFStringGetCharactersPtr` rather than allocate if possible?

Reply:
We would need to transcode into a new buffer anyway, wouldn't we?